pkg/controller/compute/v1beta1/cluster.go

```diff
@@ -0,0 +1,540 @@
+	}
+}
+
+// newDatabaseEncryptionUpdate returns a function that updates the DatabaseEncryption of a cluster.
+func newDatabaseEncryptionUpdate(in *v1beta1.DatabaseEncryption) updateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		out := &container.Cluster{}
+		gke.GenerateDatabaseEncryption(in, out)
+		update := &container.UpdateClusterRequest{
+			Update: &container.ClusterUpdate{
+				DesiredDatabaseEncryption: out.DatabaseEncryption,
+			},
+		}
+		return s.Projects.Locations.Clusters.Update(name, update).Context(ctx).Do()
+	}
+}
+
+// newLegacyAbacUpdate returns a function that updates the LegacyAbac of a cluster.
+func newLegacyAbacUpdate(in *v1beta1.LegacyAbac) updateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		out := &container.Cluster{}
+		gke.GenerateLegacyAbac(in, out)
+		update := &container.SetLegacyAbacRequest{
+			Enabled: out.LegacyAbac.Enabled,
+		}
+		return s.Projects.Locations.Clusters.SetLegacyAbac(name, update).Context(ctx).Do()
+	}
+}
+
+// newLocationsUpdate returns a function that updates the Locations of a cluster.
+func newLocationsUpdate(in []string) updateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		update := &container.UpdateClusterRequest{
+			Update: &container.ClusterUpdate{
+				DesiredLocations: in,
+			},
+		}
+		return s.Projects.Locations.Clusters.Update(name, update).Context(ctx).Do()
+	}
+}
+
+// newLoggingServiceUpdate returns a function that updates the LoggingService of a cluster.
+func newLoggingServiceUpdate(in *string) updateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		update := &container.UpdateClusterRequest{
+			Update: &container.ClusterUpdate{
+				DesiredLoggingService: gcp.StringValue(in),
+			},
+		}
+		return s.Projects.Locations.Clusters.Update(name, update).Context(ctx).Do()
+	}
+}
+
+// newMaintenancePolicyUpdate returns a function that updates the MaintenancePolicy of a cluster.
+func newMaintenancePolicyUpdate(in *v1beta1.MaintenancePolicySpec) updateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		out := &container.Cluster{}
+		gke.GenerateMaintenancePolicy(in, out)
+		update := &container.SetMaintenancePolicyRequest{
+			MaintenancePolicy: out.MaintenancePolicy,
+		}
+		return s.Projects.Locations.Clusters.SetMaintenancePolicy(name, update).Context(ctx).Do()
+	}
+}
+
+// newMasterAuthUpdate returns a function that updates the MasterAuth of a cluster.
+func newMasterAuthUpdate(in *v1beta1.MasterAuth) updateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		out := &container.Cluster{}
+		gke.GenerateMasterAuth(in, out)
+		update := &container.SetMasterAuthRequest{
+			// TODO(hasheddan): need to set Action here?
```

Comment:
Action is used to update `Username` / `Password` but we do not expose that to the user. We only expose whether we want a certificate to be generated, which I don't think can be undone so the change will be irrelevant, but you are able to attempt it, which is why it is being left in. Might be worth removing to avoid confusion though?

Reply:
I have confirmed that client `issueClientCertificate` is [unable to be updated](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters/setMasterAuth) following creation. I am going to remove the `MasterAuth` update and mark `issueClientCertificate` as immutable given that it [can not be removed](https://cloud.google.com/kubernetes-engine/docs/how-to/hardening-your-cluster#restrict_authn_methods) after it is created.